File: scripts/validate-data.mjs
#!/usr/bin/env node
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const dataDir = path.resolve(__dirname, "..", "data");

function read(file) {
  return JSON.parse(fs.readFileSync(path.join(dataDir, file), "utf-8"));
}

const cases = read("cases.json");
const orderTemplates = read("order_templates.json");
const preSurvey = read("pre_survey.json");
const postSurvey = read("post_survey.json");
const caseQuickSurvey = read("case_quick_survey.json");

const errors = [];
const warn = [];

const officialIds = cases.filter((c) => !c.isPractice).map((c) => c.id);
const defectiveIds = new Set(
  cases.filter((c) => !c.isPractice && c.defectPresent).map((c) => c.id),
);
const highRiskIds = new Set(
  cases.filter((c) => !c.isPractice && c.riskLevel === "high").map((c) => c.id),
);
const lowRiskIds = new Set(
  cases.filter((c) => !c.isPractice && c.riskLevel === "low").map((c) => c.id),
);

if (officialIds.length !== 8) {
  errors.push(`正式 case 数量应为 8，实际 ${officialIds.length}`);
}

const requiredCaseFields = [
  "id",
  "isPractice",
  "riskLevel",
  "defectPresent",
  "patientMessage",
  "chartSnapshot",
  "aiDraft",
  "guardrail",
  "goldAction",
];
for (const c of cases) {
  for (const f of requiredCaseFields) {
    if (c[f] === undefined || c[f] === null) {
      errors.push(`case ${c.id} 缺少字段 ${f}`);
    }
  }
  if (c.guardrail) {
    if (!Array.isArray(c.guardrail.factsUsed) || c.guardrail.factsUsed.length === 0) {
      errors.push(`case ${c.id} guardrail.factsUsed 为空`);
    }
    if (!c.guardrail.riskCue) {
      errors.push(`case ${c.id} guardrail.riskCue 为空`);
    }
    if (!Array.isArray(c.guardrail.checklist) || c.guardrail.checklist.length < 3) {
      errors.push(`case ${c.id} guardrail.checklist 应至少 3 条`);
    }
  }
  const allowedActions = new Set([
    "send_as_is",
    "edit_then_send",
    "discard_and_rewrite",
    "escalate",
  ]);
  if (!allowedActions.has(c.goldAction)) {
    errors.push(`case ${c.id} goldAction 不在四种动作内`);
  }
}

if (!Array.isArray(orderTemplates.templates) || orderTemplates.templates.length !== 4) {
  errors.push(`需要 4 套顺序模板，实际 ${orderTemplates.templates?.length}`);
}

for (const t of orderTemplates.templates) {
  if (t.order.length !== 8) {
    errors.push(`模板 ${t.id} 顺序长度应为 8`);
  }
  const set = new Set(t.order);
  if (set.size !== t.order.length) {
    errors.push(`模板 ${t.id} 包含重复 case`);
  }
  for (const id of t.order) {
    if (!officialIds.includes(id)) {
      errors.push(`模板 ${t.id} 包含未知 case ${id}`);
    }
  }

  let consecutiveDefective = 0;
  let maxConsecutiveDefective = 0;
  for (const id of t.order) {
    if (defectiveIds.has(id)) {
      consecutiveDefective += 1;
      maxConsecutiveDefective = Math.max(maxConsecutiveDefective, consecutiveDefective);
    } else {
      consecutiveDefective = 0;
    }
  }
  if (maxConsecutiveDefective >= 3) {
    errors.push(`模板 ${t.id} 出现 ${maxConsecutiveDefective} 个 defective case 连续`);
  }

  const lastHalf = t.order.slice(4);
  const highInLastHalf = lastHalf.filter((id) => highRiskIds.has(id)).length;
  const allHighInLast = highInLastHalf === highRiskIds.size;
  if (allHighInLast) {
    errors.push(`模板 ${t.id} 所有高风险 case 都集中在后半段`);
  }

  const firstHalf = t.order.slice(0, 4);
  const lowAccurateIds = [...lowRiskIds].filter((id) => !defectiveIds.has(id));
  const lowAccurateInFirstHalf = firstHalf.filter((id) => lowAccurateIds.includes(id)).length;
  if (lowAccurateInFirstHalf === lowAccurateIds.length && lowAccurateIds.length > 0) {
    warn.push(`模板 ${t.id} 所有 low-accurate case 都在前半段（建议打散）`);
  }
}

if (!Array.isArray(preSurvey.items) || preSurvey.items.length === 0) {
  errors.push("pre_survey.items 为空");
}
if (!Array.isArray(postSurvey.blocks) || postSurvey.blocks.length === 0) {
  errors.push("post_survey.blocks 为空");
}
if (!Array.isArray(caseQuickSurvey.items) || caseQuickSurvey.items.length !== 3) {
  errors.push("case_quick_survey 应有 3 个题目");
}

if (errors.length === 0) {
  console.log("✅ 数据校验通过");
  console.log(`  正式 case: ${officialIds.length}`);
  console.log(`  defective: ${defectiveIds.size}`);
  console.log(`  high-risk: ${highRiskIds.size}`);
  console.log(`  low-risk: ${lowRiskIds.size}`);
  console.log(`  顺序模板: ${orderTemplates.templates.length}`);
  if (warn.length > 0) {
    console.log("");
    console.log("⚠️  警告：");
    for (const w of warn) console.log("  -", w);
  }
  process.exit(0);
} else {
  console.error("❌ 数据校验失败：");
  for (const e of errors) console.error("  -", e);
  if (warn.length > 0) {
    console.error("");
    console.error("⚠️  另有警告：");
    for (const w of warn) console.error("  -", w);
  }
  process.exit(1);
}

File: backend/app/scripts/validate_data.py
"""
================================================================================
文件作用：★ 题库的体检工具 —— 改完题目一定要跑一遍再提交
================================================================================

    cd backend
    PYTHONIOENCODING=utf-8 python -m app.scripts.validate_data

它检查的是那些「程序不会报错、但会毁掉研究」的问题：

  * 正式题不是 8 道
  * goldAction 写了四种动作之外的值（比如拼错成 "send_as_it"）
  * 某道题没写次优答案，导致判分过于严苛
  * ★ 顺序模板里，唯一那道「AI 草稿没出错」的题排到了后半段

  最后这一条最要命：如果医生前几题看到的 AI 草稿全都有错，
  他会很快形成「这个 AI 总是出错」的预期，之后所有判断都带着这个偏见。
  所以那道"没错"的题必须早点出现，把基线拉正。

发现任何错误就返回非零退出码，CI 会因此失败，提交不上去。

--------------------------------------------------------------------------------
本文件的代码块：
--------------------------------------------------------------------------------
  第 1 块  main()   全部检查都在这一个函数里，按关卡顺序往下走
================================================================================
"""

from __future__ import annotations

import sys

from app.schemas.case import CaseRaw
from app.schemas.common import SelectedAction
from app.scripts.data_loader import (
    load_case_quick_survey,
    load_cases,
    load_order_templates,
    load_post_survey,
    load_pre_survey,
)


# ── 第 1 块：全部检查 ────────────────────────────────────────────────────
def main() -> int:
    # errors = 必须修的（会让脚本返回非零退出码）
    # warnings = 提醒一下，不阻塞
    errors: list[str] = []
    warnings: list[str] = []

    # ---- 第 1 关：每道题的字段格式对不对 ----
    cases_raw = load_cases()
    cases: list[CaseRaw] = []
    for raw in cases_raw:
        try:
            cases.append(CaseRaw.model_validate(raw))
        except Exception as e:
            # 收集错误而不是直接抛出，这样一次能报出所有问题，
            # 不用改一个跑一次。
            errors.append(f"case {raw.get('id')}: schema invalid -> {e}")

    # ---- 第 2 关：题目数量 ----
    formal = [c for c in cases if not c.isPractice]
    if len(formal) != 8:
        errors.append(f"正式 case 数量应为 8，实际 {len(formal)}")

    # ---- 第 3 关：每道题的内容完整性 ----
    allowed_actions = {a.value for a in SelectedAction}
    for c in cases:
        if c.goldAction not in allowed_actions:
            errors.append(f"case {c.id} goldAction 不在四种动作内: {c.goldAction}")
        if not c.guardrail.factsUsed:
            errors.append(f"case {c.id} guardrail.factsUsed 为空")
        if not c.guardrail.riskCue:
            errors.append(f"case {c.id} guardrail.riskCue 为空")
        if len(c.guardrail.checklist) < 3:
            errors.append(f"case {c.id} guardrail.checklist 应至少 3 条")

    # ---- 第 4 关：题目顺序模板的约束 ----
    # 先把题目按属性分组，下面检查顺序模板时要用
    formal_ids = {c.id for c in formal}
    defective_ids = {c.id for c in formal if c.defectPresent}      # AI 有错的题（7 道）
    non_defective_ids = formal_ids - defective_ids                 # AI 没错的题（1 道）
    high_risk_ids = {c.id for c in formal if c.riskLevel == "high"}
    low_risk_ids = {c.id for c in formal if c.riskLevel == "low"}

    templates = load_order_templates().get("templates", [])
    if len(templates) != 4:
        errors.append(f"需要 4 套顺序模板，实际 {len(templates)}")
    for t in templates:
        order = t["order"]
        if len(order) != 8:
            errors.append(f"模板 {t['id']} 顺序长度应为 8")
        if len(set(order)) != len(order):
            errors.append(f"模板 {t['id']} 包含重复 case")
        for cid in order:
            if cid not in formal_ids:
                errors.append(f"模板 {t['id']} 包含未知 case {cid}")

        first_half = order[:4]
        last_half = order[4:]

        # V3 加难版：仅保留 1 个 non-defect 校准 case（case_01）时，要求它必须出现
        # 在前半段，让受试者在前期看到 AI 至少有一次写对，避免「AI 必有错」的预期。
        if non_defective_ids:
            if not (non_defective_ids & set(first_half)):
                errors.append(
                    f"模板 {t['id']} 前半段未包含任何 non-defect case；"
                    f"trust 校准要求至少 1 个 non-defect 在位置 1-4"
                )

        # 多元素集合时才检查「不全在某半段」（单 case 集合下该约束无意义）
        if len(high_risk_ids) >= 2:
            if len([c for c in last_half if c in high_risk_ids]) == len(high_risk_ids):
                errors.append(f"模板 {t['id']} 所有高风险 case 都集中在后半段")
        low_accurate = low_risk_ids - defective_ids
        if len(low_accurate) >= 2:
            if len([c for c in first_half if c in low_accurate]) == len(low_accurate):
                warnings.append(f"模板 {t['id']} 所有 low-accurate case 都在前半段（建议打散）")

    pre = load_pre_survey()
    if not pre.get("items"):
        errors.append("pre_survey.items 为空")
    post = load_post_survey()
    if not post.get("blocks"):
        errors.append("post_survey.blocks 为空")
    quick = load_case_quick_survey()
    if not quick.get("items") or len(quick["items"]) != 2:
        errors.append("case_quick_survey 应有 2 个题目")

    if not errors:
        print("✅ 数据校验通过")
        print(f"  正式 case: {len(formal)}")
        print(f"  defective: {len(defective_ids)}")
        print(f"  high-risk: {len(high_risk_ids)}")
        print(f"  low-risk: {len(low_risk_ids)}")
        print(f"  顺序模板: {len(templates)}")
        if warnings:
            print("\n⚠️  警告：")
            for w in warnings:
                print("  -", w)
        return 0

    print("❌ 数据校验失败：")
    for e in errors:
        print("  -", e)
    if warnings:
        print("\n⚠️  另有警告：")
        for w in warnings:
            print("  -", w)
    return 1


if __name__ == "__main__":
    sys.exit(main())
